### ISSUE TITLE
MyCaffeControl.Clone() Null Exception

### ISSUE THREAD

**Issue by YanYas** [MyCaffe/MyCaffe#15]
Hi again,

I've been playing with ImageClassificationBarebones example. I would like to use the MyCaffeControl object I created, take the network that has been trained with the Solver, and **Run()** the model with a SimpleDatum created from one of the MNIST images.
Where I'm confused is by how I set the Phase of the Net to 'RUN'

I thought of copying or Cloning the MyCaffeControl or the Net itself. Cloning the MyCaffeControl currently returns a Null object error and there isn't an option for Cloning the Net(), only the NetParameters. The cloning the parameters means I have to copy the weights of the model to the CPU and back to GPU.

What seems to happen at the moment is when I call  Clone on MyCaffeControl, internally it creates the internal objects correctly until it gets to caffe.Load().
One of the arguments of this function is a ProjectEx m_project which in my case is Null because I'm not using a database.

is there also a tidy way to share the Net() for Running?

All the best,

Yy

**Comment by ZoroDerVonCodier** [comment]
Hi @YanYas ,

We will look into the Clone null issue. We have added to the ImageClassificationBareBones sample with a sub-sample that shows how to run a trained model (weights) on a set of images loaded from file.  Check out the new btnTestTrainedWeights_Click method in https://github.com/MyCaffe/MyCaffe-Samples/blob/master/ImageClassification/ImageClassificationBareBones/MainForm.cs which shows how to use the MyCaffeControl.LoadToRun method which loads a set of already trained weights.

For more information on LoadToRun, see https://www.signalpop.com/onlinehelp/mycaffe/html/class_my_caffe_1_1_my_caffe_control.html#ac0567e4ee39af7cc026a3880dc31e0ac

Hope this helps!

**Comment by ZoroDerVonCodier** [comment]
Hi @YanYas ,

Currently, the LoadLite does not create the internal Run Network so at this point, you will need to use the 'LoadToRun' method.  We will look into adding functionality to LoadLite in a future release that allows you to optionally load the Run network along with the internal Training and Test networks.  The Load method (which uses the SQL database) has this functionality and loads all three when called.

We have been able to repro that this causes the null pointer issue in Clone and will also cause a similar issue in GetWeights().  We will fixe this in the next release.

In the mean time, the sample updates noted above provide a work around.

Hope this helps!

**Comment by ZoroDerVonCodier** [comment]
@YanYas ,
Could you provide a short sample of the code that causes a null when calling Clone?

**Comment by ZoroDerVonCodier** [comment]
@YanYas ,
No need for the sample, we are able to repro the issue which ironically relates to the internal Run net no being created by the LoadLite. This also impacts several other methods like GetWeights().  We will get this fixed in the next release.  In the mean-time the sample updates show how to

**Comment by ZoroDerVonCodier** [comment]
Hi @YanYas ,
Good catches - we have fixed the issues where LoadLite now creates the internal Run net, we fixed the issues in Clone and other methods that depended on the internal Run net are now fixed as well.

These will be in the next release.  If you are building locally and would like to test in the sample, we have added to that as well - just define the VER_10_2_160 Conditional compilation symbol.

Hope this helps!

**Comment by ZoroDerVonCodier** [comment]
Hi @YanYas ,

We just added a new sample to the ImageClassificationBareBones project that shows how to use the new ImageDataLayer that we added to the GitHub project.  The ImageDataLayer handles loading a dataset of images from file without using SQL.  The MyCaffe Test App will now create the file_list.txt file when exporting the MNIST dataset - just use the path and file name as the source and you are ready to do.  The sample shows how to do this.

Hope this helps!
